### PR TITLE
Including a schemeURI with a nameIdentifier that is a URL will concat…

### LINKED ISSUE #136: Including a schemeURI with a nameIdentifier that is a URL will concatenate the schemeURI and the nameIdentifier
### Describe the bug
SchemeURIs and nameIdentifiers that are URLs are unexpectedly concatenated if a schemeURI is included. 

#### Expected Behaviour
The entered nameIdentifier should either remain untouched or the system should intelligently avoid incorrect concatenation of schemeURIs and nameIdentifiers. 

#### Current Behaviour

Providing XML . . . 

`<contributor contributorType="Sponsor">
<contributorName nameType="Organizational">Gump South Pacific Research Station (moorea.berkeley.edu)</contributorName>
<nameIdentifier nameIdentifierScheme="ROR" schemeURI="https://ror.org/">https://ror.org/04sk0et52</nameIdentifier>
</contributor>
</contributors>`

. . . generates a nameIdentifier where the schemeURI and the nameIdentifier are concatenated:

`https://ror.org/https://ror.org/04sk0et52`

#### Hypothesis

The code concatenates schemeURIs and nameIdentifiers if a schemeURI is present and the identifier is not an ORCID:
https://github.com/datacite/bolognese/blob/712184f1c82def2b0c500bb4328bae6b7dc71dee/lib/bolognese/author_utils.rb#L33

#### Possible Implementation

The code here could adapt to URL nameIdentifiers rather than simply concatenating schemeURIs and nameIdentifiers: 
https://github.com/datacite/bolognese/blob/712184f1c82def2b0c500bb4328bae6b7dc71dee/lib/bolognese/author_utils.rb#L33
Similarly to how URL affiliationIdentifiers are handled:
https://github.com/datacite/bolognese/blob/712184f1c82def2b0c500bb4328bae6b7dc71dee/lib/bolognese/author_utils.rb#L152

### PR DESCRIPTION
**Pull request by codycooperross**
…enate the schemeURI and the nameIdentifier

Fixes #136

## Purpose

When a nameIdentifier property contains both a schemeURI attribute and a nameIdentifier in the form of a URL, bolognese concatenates the schemeURI and nameIdentifier, resulting in incorrectly normalized nameIdentifiers like `https://ror.org/https://ror.org/04p8xrf95 `. This does not affect ORCIDs, which have specific normalization logic applied. 

closes: #136

## Approach

Rather than simply concatenating schemeURIs and nameIdentifiers when a schemeURI is present, the code now adopts the logic from the affiliationIdentifier code:

https://github.com/datacite/bolognese/blob/5a01ebd48784fa837db351405a8153e1bbb5c18d/lib/bolognese/author_utils.rb#L147

This only concatenates the schemeURI if "https://" is not present in the nameIdentifier and a schemeURI is present. It also performs some additional URL normalization found here:

https://github.com/datacite/bolognese/blob/5a01ebd48784fa837db351405a8153e1bbb5c18d/lib/bolognese/utils.rb#L610

#### Open Questions and Pre-Merge TODOs

Unfortunately, applying this logic to, for example, ISNIs can be problematic. As it currently exists, the implementation fails a test where nameIdentifier `0000 0003 5752 6882` with schemeURI `http://isni.org/isni/` is normalized to `http://isni.org/isni/0000%200003%205752%206882` (with the spaces escaped) when the test expects `http://isni.org/isni/0000 0003 5752 6882`. I would argue that neither the original test nor the fix produce correct results, since the ISNI is correct in the original expression `0000 0003 5752 6882`. 

The available options seem to be:

- Keep the current fix but apply it only to RORs and add ROR-specific logics, e.g. always assume the schemeURI is `https://ror.org/` when the nameIdentifierScheme is `ROR`.  For other identifier schemes, simply transcribe the schemeURI and nameIdentifier as entered. 
- Create additional logics for other identifiers like ISNIs.
- Make no attempt at normalization and leave nameIdentifiers and schemeURIs as entered if they are not ORCIDs. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

- [ ] New feature (non-breaking change which adds functionality)

- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Reviewer, please remember our [guidelines](https://datacite.atlassian.net/wiki/spaces/TEC/pages/1168375809/Pull+Request+Guidelines):

- Be humble in the language and feedback you give, ask don't tell.
- Consider using positive language as opposed to neutral when offering feedback. This is to avoid the negative bias that can occur with neutral language appearing negative.
- Offer suggestions on how to improve code e.g. simplification or expanding clarity.
- Ensure you give reasons for the changes you are proposing.

### CODE REVIEW

File: lib/bolognese/author_utils.rb [hunk 1]
```diff
@@ -36,9 +36,15 @@ def get_one_author(author)
             "schemeUri" => "https://orcid.org",
             "nameIdentifierScheme" => "ORCID" }.compact
         elsif ni["schemeURI"].present?
+          name_identifier = nil
+          if ni["__content__"].present?
+            name_identifier = ni["__content__"].to_s.strip
+            schemeURI = ni["schemeURI"].end_with?("/") ? ni["schemeURI"] : ni["schemeURI"] + "/"
```
Comment (kjgarza) [inline lib/bolognese/author_utils.rb:42]: maybe abstract this out to its own method. it would make the code easier to read next time around.

**Review by richardhallett — APPROVED** [review]
lgtm

**Comment by kjgarza** [comment]
we just found out that 
image

**Comment by richardhallett** [comment]
Having looked at this again, maybe go with option 1:
`Keep the current fix but apply it only to RORs and add ROR-specific logics, e.g. always assume the schemeURI is https://ror.org/ when the nameIdentifierScheme is ROR. For other identifier schemes, simply transcribe the schemeURI and nameIdentifier as entered.`

i.e. lets just handle ROR as the main issue here for now and not try and be clever elsewhere.

**Review by richardhallett — APPROVED** [review]

**Review by digitaldogsbody — APPROVED** [review]
Good fix :)

**Review by jrhoads — APPROVED** [review]

**Pull request merged**